src/Concurrency.h

```diff
@@ -0,0 +1,144 @@
+/* This file is part of the CARTA Image Viewer: https://github.com/CARTAvis/carta-backend
+   Copyright 2018, 2019, 2020, 2021 Academia Sinica Institute of Astronomy and Astrophysics (ASIAA),
+   Associated Universities, Inc. (AUI) and the Inter-University Institute for Data Intensive Astronomy (IDIA)
+   SPDX-License-Identifier: GPL-3.0-or-later
+*/
+
+#pragma once
+
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <shared_mutex>
+
+#define __USE_TBB__ 0
+
+template <class T>
+class concurrent_queue {
+public:
+    concurrent_queue<T>() {}
+    ~concurrent_queue<T>() {
+        clear();
+    }
+    void push(T elt) {
+        _mtx.lock();
+        _q.push_back(elt);
+        _mtx.unlock();
+    }
+    bool try_pop(T& elt) {
+        bool ret = false;
+        _mtx.lock();
+        if (!_q.empty()) {
+            ret = true;
+            elt = _q.front();
+            _q.pop_front();
+        }
+        _mtx.unlock();
+        return ret;
+    }
+    void clear() {
+        _mtx.lock();
+        while (!_q.empty()) {
+            _q.pop_front();
+        }
+        _mtx.unlock();
+    }
+
+private:
+    std::list<T> _q;
+    std::mutex _mtx;
+};
+
+class queuing_rw_mutex {
+public:
+    queuing_rw_mutex() {
+        _reader_count = 0;
+        _writer_count = 0;
+    }
+    ~queuing_rw_mutex() {
+        while (!_mtx_list.empty()) {
+            delete _mtx_list.front();
+            _mtx_list.pop_front();
+        }
+        _readers_cv.notify_all();
+    }
+    void reader_enter() {
+        std::unique_lock<std::mutex> lock(_mtx);
+        if (_writer_count > 0) {
+            _readers_cv.wait(lock);
+        } else {
+            ++_reader_count;
```

Comment:
Should this be incremented conditionally here? It's decremented unconditionally in `reader_leave`.